docs/language-spec.md

```diff
@@ -1334,6 +1418,534 @@ does not currently support additional path components in the type URL and does n
+    oneof id {                                // message foo.bar.Message
+      string name = 1;                        // message foo.bar.Message
+      uint64 num = 2;                         // message foo.bar.Message
+    }                                         //
+    message NestedMessage {                   // message foo.bar.Message
+      extend google.protobuf.MessageOptions { //
+        string fizz = 49999;                  // message foo.bar.Message.NestedMessage
+      }                                       //
+      option (NestedMessage.fizz) = "buzz";   //
+      enum Kind {                             // message foo.bar.Message.NestedMessage
+        NULL = 0;                             // enum foo.bar.Message.NestedMessage.Kind
+        PRIMARY = 1;                          // enum foo.bar.Message.NestedMessage.Kind
+        SECONDARY = 2;                        // enum foo.bar.Message.NestedMessage.Kind
+      }                                       //
+      Kind kind = 1;                          // message foo.bar.Message.NestedMessage
+    }                                         //
+    NestedMessage extra = 3;                  // message foo.bar.Message
+    extensions 1 to 100;                      // message foo.bar.Message
+}                                             //
+                                              //
+enum Unit {                                   // file "example.proto"
+  VOID = 0;                                   // enum foo.bar.Unit
+}                                             //
+                                              //
+service FooService {                          // file "example.proto"
+  rpc Bar(Message) returns (Message);         // service foo.bar.FooService
+}                                             //
+```
+The file itself has no parent element: it is the root of the hierarchy.
+
+The algorithm to resolve a feature for a particular element looks like so:
+1. Examine the element's options.
+   * If available, examine the option message's `features` field.
+     * If present, query for the particular feature field in question.
+       * If present, its value is the resolved feature value. **Done.**
+2. If the element is a file (so there is no parent), the resolved
+   feature value is the default value for the file's edition. **Done.**
+3. Get the element's parent.
+4. Go back to step 1, repeating the algorithm, but substituting parent
+   for element.
+
+Computing the default value of a feature for a particular edition is discussed
+in the [next section](#feature-defaults).
+
+An alternative to the above algorithm is to pre-compute the resolved values for all
+features, for all elements. This utilizes a logical "merge" operation, to combine two
+Protobuf messages. It also traverses the hierarchy in the opposite direction, top-down
+instead of bottom-up. The algorithm for this approach looks like so:
```

Comment:
```suggestion
features, for all elements. This utilizes a logical ["merge" operation](#merging-protobuf-messages)
to combine two Protobuf messages. It also traverses the hierarchy in the opposite
direction, top-down instead of bottom-up. The algorithm for this approach looks
like so:
```